### PR TITLE
Add worldstate heal mechanism 

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

## PR description

This PR add a heal  mechanism fo the worldtstate in case of inconsistency (unable to trie node). To fix this, we start a quick heal of the worldstate automatically and once the fix is done we restart the block import.

After the detection of an invalid path
- we delete this path to force the healing of this part.
- then we delete the trielogs. 
- we select a pivot block (before of after the heal)
- we move the blokchain to this pivot block (rewind or download the missing blocks) 
- then we launch a worldstate heal. 

This feature can also heal a node that has been inconsistent for a long time, but it will take longer because there will be more nodes to heal.  With this PR the healing will be done as soon as the problem is detected so there will not a lot to heal and it will be fast

## Performed tests
 - Trigger multiple inconsistencies to fix  (passed)
 - Fixed a node that has been inconsistent for a long time (passed)
 - Run a snapsync from scratch on goerli (passed)
 - Run a checkpoint sync from scratch on goelri (passed)
 - Run a checkpoint sync from scratch on main **(in progress)**
 - Run a validator teku+besu on goerli (passed)
 - Profile performance on this node to avoid perf regression (passed)

## Fixed Issue(s)

#4379 
#4785 
#4768

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Comment by matkt** [comment]
Initial PR is here https://github.com/hyperledger/besu/pull/4972

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/FinalBlockConfirmation.java [hunk 1]
```diff
@@ -44,7 +44,9 @@ static FinalBlockConfirmation genesisConfirmation(final MutableBlockchain blockc
   }
 
   static FinalBlockConfirmation ancestorConfirmation(final MutableBlockchain blockchain) {
-    return firstHeader -> blockchain.contains(firstHeader.getParentHash());
+    return firstHeader ->
+        blockchain.contains(firstHeader.getParentHash())
+            && blockchain.getChainHeadBlockNumber() + 1 >= firstHeader.getNumber();
```
Comment (siladu) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/FinalBlockConfirmation.java:49]: @matkt This appears to be different from the other PR: https://github.com/hyperledger/besu/pull/4972/files#diff-0361d2ff12cd43f507bf88a3cb4d781a537ca765329e0811f60a9507527cbe13R47

Comment (siladu) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/FinalBlockConfirmation.java:49]: looks fine though, I will assume it's an extra safety check.

**Review by siladu — COMMENTED** [review]
Looks like be an accurate recreation of https://github.com/hyperledger/besu/pull/4972 except for the minor change I mentioned in the comment

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/FinalBlockConfirmation.java (hunk 1 above)
Comment (siladu) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/FinalBlockConfirmation.java:49]: ...ah just caught up with the new comments from the older PR :)
https://github.com/hyperledger/besu/pull/4972#discussion_r1097020981

**Review by siladu — APPROVED** [review]
LGTM as someone with not much in-depth bonsai knowledge. Ideally someone else with more experience in this area will approve :)

**Review by macfarla — COMMENTED** [review]
LGTM (but I am also no bonsai expert)
Needs a changelog entry

**Review by macfarla — APPROVED** [review]
LGTM

**Pull request merged**